src/bookshelf/src/bksw_wm_sql_object.erl

```diff
@@ -0,0 +1,336 @@
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(bksw_wm_sql_object).
+
+-include_lib("mixer/include/mixer.hrl").
+-mixin([{bksw_wm_base, [init/1,
+                        is_authorized/2,
+                        finish_request/2,
+                        service_available/2]}]).
+
+%% Webmachine callbacks
+-export([allowed_methods/2,
+         content_types_accepted/2,
+         content_types_provided/2,
+         delete_resource/2,
+         generate_etag/2,
+         last_modified/2,
+         resource_exists/2,
+
+         %% Override
+         validate_content_checksum/2,
+
+         %% Resource helpers
+         download/2,
+         upload/2]).
+
+
+-include_lib("webmachine/include/webmachine.hrl").
+-include("internal.hrl").
+
+%%===================================================================
+%% Public API
+%%===================================================================
+
+%% By default, if wm sees a 'content-md5' header, it will read the request body to compute
+%% the md5 and compare to the header value. A 400 will then be returned automagically by wm
+%% if the digests do not match. Since we wish to read request bodies in a streaming fashion,
+%% we need to handle our own checksum validation. Using wm's default would mean having a
+%% full copy of the request body buffered into the request process state. So we define this
+%% resource callback to blindly say the content is valid and then do the verification in the
+%% upload/2 flow.
+validate_content_checksum(Rq, Ctx) ->
+    {true, Rq, Ctx}.
+
+allowed_methods(Rq, Ctx) ->
+    {['HEAD', 'GET', 'PUT', 'DELETE'], Rq, Ctx}.
+
+content_types_provided(Rq, Ctx) ->
+    CType =
+    case wrq:get_req_header("accept", Rq) of
+        undefined ->
+            "application/octet-stream";
+        "*/*" ->
+            "application/octet-stream";
+        C ->
+            C
+    end,
+    {[{CType, download}], Rq, Ctx}.
+
+content_types_accepted(Rq, Ctx) ->
+    CT = case wrq:get_req_header("content-type", Rq) of
+             undefined ->
+                 "application/octet-stream";
+             X ->
+                 X
+         end,
+    {MT, _Params} = webmachine_util:media_type_to_detail(CT),
+    {[{MT, upload}], Rq, Ctx}.
+
+resource_exists(Rq0, Ctx) ->
+    case wrq:method(Rq0) of
+        %% Buckets always exist for writes since we create them on the fly
+        'PUT' ->
+            case fetch_entry_md(Rq0, Ctx) of
+                {error, _Error} ->
+                    {false, Rq0, Ctx};
+                {not_found, #context{} = Ctx1 } ->
```

Comment:
I'm also confused why we're returning `true` for `resource_exists` when the resource doesn't actually exist. Perhaps we should talk through this because reading the [webmachine state machine](https://raw.githubusercontent.com/webmachine/webmachine/master/docs/http-headers-status-v3.png) isn't really helpful to me right now.